buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java

```diff
@@ -0,0 +1,196 @@
+                variant.getAttributes().attribute(ArtifactAttributes.ARTIFACT_FORMAT, ArtifactTypeDefinition.DIRECTORY_TYPE);
+                variant.artifactsProvider(() -> Collections.singletonList(new DirectoryPublishArtifact(explodedArchiveTask)));
+            });
+        });
+        project.getExtensions().add("distribution_archives", container);
+    }
+
+    private void configureGeneralTaskDefaults(Project project) {
+        // common config across all copy / archive tasks
+        project.getTasks().withType(AbstractCopyTask.class).configureEach(t -> {
+            t.dependsOn(project.getTasks().withType(EmptyDirTask.class));
+            t.setIncludeEmptyDirs(true);
+            t.setDirMode(0755);
+            t.setFileMode(0644);
+        });
+
+        // common config across all archives
+        project.getTasks().withType(AbstractArchiveTask.class).configureEach(t -> {
+            String subdir = buildTaskToSubprojectName(t.getName());
+            t.getDestinationDirectory().set(project.file(subdir + "/build/distributions"));
+            t.getArchiveBaseName().set(subdir.contains("oss") ? "elasticsearch-oss" : "elasticsearch");
+        });
+    }
+
+    private void configureTarDefaults(Project project) {
+        // common config across all tars
+        project.getTasks().withType(SymbolicLinkPreservingTar.class).configureEach(t -> {
+            t.getArchiveExtension().set("tar.gz");
+            t.setCompression(Compression.GZIP);
+        });
+    }
+
+    private void registerEmptyDirectoryTasks(Project project) {
+        // CopySpec does not make it easy to create an empty directory so we
+        // create the directory that we want, and then point CopySpec to its
+        // parent to copy to the root of the distribution
+        File logsDir = new File(project.getBuildDir(), "logs-hack/logs");
+        project.getExtensions().add("logsDir", new File(project.getBuildDir(), "logs-hack/logs"));
+        project.getTasks().register("createLogsDir", EmptyDirTask.class, t -> {
+            t.setDir(logsDir);
+            t.setDirMode(0755);
+        });
+
+        File pluginsDir = new File(project.getBuildDir(), "plugins-hack/plugins");
+        project.getExtensions().add("pluginsDir", pluginsDir);
+        project.getTasks().register("createPluginsDir", EmptyDirTask.class, t -> {
+            t.setDir(pluginsDir);
+            t.setDirMode(0755);
+        });
+
+        File jvmOptionsDir = new File(project.getBuildDir(), "jvm-options-hack/jvm.options.d");
+        project.getExtensions().add("jvmOptionsDir", jvmOptionsDir);
+        project.getTasks().register("createJvmOptionsDir", EmptyDirTask.class, t -> {
+            t.setDir(jvmOptionsDir);
+            t.setDirMode(0750);
+        });
+    }
+
+    private String buildTaskToSubprojectName(String taskName) {
+        return taskName.substring("build".length()).replaceAll("[A-Z]", "-$0").toLowerCase().substring(1);
```

Comment:
I'm not sure we need the "build" prefix convention here for container entries. The DSL is meant to describe a distribution, the the task that eventually constructs it. The latter is an implementation detail handled by the plugin. I think it makes more sense to create a `darwinTar` and then the plugin incidentally creates a `buildDarwinTar` task.

Reply:
fixed